Pages Plugins (#4027)
cloudflare/cloudflare-docs: content/pages/platform/functions/plugins/cloudflare-access.md

---
pcx-content-type: concept
title: Cloudflare Access
weight: 1
---

# Cloudflare Access Pages Plugin

> This Plugin is a middleware to validate Cloudflare Access JWT assertions. It also includes an API to lookup additional information about a given user's JWT.

Reviewer: ```suggestion
The Cloudflare Access Pages Plugin is a middleware to validate Cloudflare Access JWT assertions. It also includes an API to lookup additional information about a given user's JWT.
```

## Installation

```sh
npm install @cloudflare/pages-plugin-cloudflare-access
```

## Usage

```typescript
---
filename: functions/_middleware.ts
---
import cloudflareAccessPlugin from "@cloudflare/pages-plugin-cloudflare-access";

export const onRequest: PagesFunction = cloudflareAccessPlugin({
  domain: "https://test.cloudflareaccess.com",
  aud: "4714c1358e65fe4b408ad6d432a5f878f08194bdb4752441fd56faefa9b2b6f2",
});
```

> The Plugin takes an object with two properties: the `domain` of your Cloudflare Access account, and the policy `aud` (audience) to validate against. Any requests which fail validation will be returned a 403.

Reviewer: ```suggestion
The Plugin takes an object with two properties: the `domain` of your Cloudflare Access account, and the policy `aud` (audience) to validate against. Any requests which fail validation will be returned a `403` status code.
```

> ### Accessing the JWT payload

Reviewer: ```suggestion
### Access the JWT payload
```

> If you need to use the JWT payload in your application (e.g. you need the user's email address), this Plugin will make this available for you at `data.cloudflareAccess.JWT.payload`.

Reviewer: ```suggestion
If you need to use the JWT payload in your application (for example, you need the user's email address), this Plugin will make this available for you at `data.cloudflareAccess.JWT.payload`.
```

For example:

```typescript
---
filename: functions/greet.ts
---
import type { PluginData } from "@cloudflare/pages-plugin-cloudflare-access";

export const onRequest: PagesFunction<unknown, any, PluginData> = async ({
  data,
}) => {
  return new Response(
    `Hello, ${data.cloudflareAccess.JWT.payload.email || "service user"}!`
  );
};
```

> The [entire JWT payload](https://developers.cloudflare.com/cloudflare-one/identity/users/validating-json/#payload) will be made available on `data.cloudflareAccess.JWT.payload`. Be aware that the fields available differ between identity authorizations (e.g. a user in a browser) and non-identity authorizations (e.g. a service token).

Reviewer: ```suggestion
The [entire JWT payload](https://developers.cloudflare.com/cloudflare-one/identity/users/validating-json/#payload) will be made available on `data.cloudflareAccess.JWT.payload`. Be aware that the fields available differ between identity authorizations (for example, a user in a browser) and non-identity authorizations (for example, a service token).
```

### Look up identity

> In order to get more information about a given user's identity, you can use the provided `getIdentity` API function:

Reviewer: ```suggestion
In order to get more information about a given user's identity, use the provided `getIdentity` API function:
```

    domain: "https://test.cloudflareaccess.com",
  });

  return new Response(`Hello, ${identity.name || "service user"}!`);
};
```

The `getIdentity` function takes an object with two properties: a `jwt` string, and a `domain` string. It returns a `Promise` of [the object returned by the `/cdn-cgi/accesss/get-identity` endpoint](https://developers.cloudflare.com/cloudflare-one/identity/users/validating-json/#groups-within-a-jwt). This is particularly useful if you want to use a user's group membership for something like application permissions.

For convience, this same information can be fetched for the current request's JWT with the `data.cloudflareAccess.JWT.getIdentity` function, (assuming you have already validated the request with the Plugin as above):

```typescript
---
filename: functions/greet.ts
---
import type { PluginData } from "@cloudflare/pages-plugin-cloudflare-access";

export const onRequest: PagesFunction<unknown, any, PluginData> = async ({
  data,
}) => {
  const identity = await data.cloudflareAccess.JWT.getIdentity();

  return new Response(`Hello, ${identity.name || "service user"}!`);
};
```

### Login and logout URLs

> If you want to force a login or logout, you can use these utility functions to generate URLs and redirect a user:

Reviewer: ```suggestion
If you want to force a login or logout, use these utility functions to generate URLs and redirect a user:
```
